File: kudos-ao3-reader/Services/WorkAvailabilitySweep.swift
import Foundation
import OSLog
import SwiftData

/// Checks a whole library for works their site has deleted, so the **Last Copy** badge
/// reflects reality rather than only what an import happened to touch.
///
/// ## Manual on purpose
///
/// This never runs by itself. Asking AO3 hundreds of questions to learn that nothing
/// changed is not a polite thing for an app to do on a timer, and Kudos has no way to
/// ask "what changed since yesterday?" — AO3 exposes no such feed, so the only
/// available implementation is one request per work. Until there is a cheaper signal,
/// the user decides when that cost is worth paying.
///
/// `AO3RequestCoordinator` bounds *concurrency* (3 at once) but does not rate-limit, so
/// a sweep left to its own devices would fire a sustained burst. This runs **serially
/// with a deliberate gap** instead: availability is maintenance, not something anyone is
/// waiting on, so there is nothing to gain from speed and a great deal to lose from
/// looking like a scraper.
///
/// Politeness measures, all of them deliberate:
/// - one request at a time, `requestSpacing` apart (≈0.7/second);
/// - works checked within `recheckInterval` are skipped, so a second run does not
///   re-ask the same questions;
/// - oldest-checked first, so repeat runs cover the library over time instead of
///   re-walking the same prefix;
/// - a per-run `limit`, with whatever it did not reach **reported** rather than
///   silently dropped;
/// - cancellable between requests, and cancellation keeps the work already done.
@MainActor
enum WorkAvailabilitySweep {
    /// Don't re-ask about a work checked more recently than this. A week: deletions are
    /// rare, and a user who runs the sweep twice in an afternoon should send no traffic
    /// the second time.
    static let recheckInterval: TimeInterval = 7 * 24 * 3600

    /// Gap between requests. 1.5s is slower than a person clicking through pages, which
    /// is the bar this should stay under.
    static let requestSpacing: Duration = .milliseconds(1500)

    /// Ceiling for one run, so a 2,000-work library cannot turn into a 50-minute
    /// session by accident. What is left over is reported and picked up next time.
    static let defaultLimit = 150

    struct Summary: Equatable {
        var checked = 0
        var stillAvailable = 0
        var nowUnavailable = 0
        /// Skipped because they were checked within `recheckInterval`.
        var skippedRecent = 0
        /// In scope but beyond this run's limit, or not reached before cancellation.
        var remaining = 0
        var cancelled = false

        /// Works that can never be checked: no AO3 identity, so no site to ask. Counted
        /// so the summary can say why a number is smaller than the library.
        var unverifiable = 0
    }

    /// Works this sweep would actually contact AO3 about, oldest check first.
    static func pending(in context: ModelContext, now: Date = Date()) -> [SavedWork] {
        allAO3Works(in: context)
            .filter { work in
                guard let last = work.lastAvailabilityCheck else { return true }
                return now.timeIntervalSince(last) >= recheckInterval
            }
            // nil (never checked) sorts first: those are the works we know least about.
            .sorted { ($0.lastAvailabilityCheck ?? .distantPast) < ($1.lastAvailabilityCheck ?? .distantPast) }
    }

    /// Every work with an AO3 identity. The rest cannot be checked at all — there is no
    /// native path to fanfiction.net or Wattpad — and are counted, not contacted.
    private static func allAO3Works(in context: ModelContext) -> [SavedWork] {
        ((try? context.fetch(FetchDescriptor<SavedWork>())) ?? [])
            .filter { !$0.isPendingDeletion && $0.origin.supportsLiveLookup }
    }

    /// Runs the sweep. Never throws: a work that cannot be checked is left as it was.
    ///
    /// `verify` is injectable so tests exercise the pacing, limit and summary logic
    /// without a network — the same seam `WorkAvailability` and
    /// `ReadingQueueService.preserve` use.
    static func run(
        in context: ModelContext,
        limit: Int = defaultLimit,
        spacing: Duration = requestSpacing,
        now: Date = Date(),
        progress: (_ completed: Int, _ total: Int) -> Void = { _, _ in },
        verify: (SavedWork) async -> Void = { await WorkAvailability.verify($0, in: $0.modelContext!) }
    ) async -> Summary {
        var summary = Summary()
        let all = (try? context.fetch(FetchDescriptor<SavedWork>())) ?? []
        summary.unverifiable = all.filter { !$0.isPendingDeletion && !$0.origin.supportsLiveLookup }.count

        let queue = pending(in: context, now: now)
        summary.skippedRecent = allAO3Works(in: context).count - queue.count

        let batch = queue.prefix(limit)
        summary.remaining = queue.count - batch.count

        for (index, work) in batch.enumerated() {
            if Task.isCancelled {
                summary.cancelled = true
                summary.remaining += batch.count - index
                break
            }
            // Spacing goes *before* each request after the first, so a cancelled sweep
            // never leaves a pointless trailing sleep, and the first check feels
            // immediate.
            if index > 0 {
                do {
                    try await Task.sleep(for: spacing)
                } catch {
                    summary.cancelled = true
                    summary.remaining += batch.count - index
                    break
                }
            }

            let wasUnavailable = work.ao3Unavailable
            await verify(work)
            summary.checked += 1
            // Read the flag rather than trusting the call: an inconclusive check (a
            // timeout) deliberately records nothing, and must not be counted either way.
            if work.lastAvailabilityCheck != nil || wasUnavailable != work.ao3Unavailable {
                if work.ao3Unavailable {
                    summary.nowUnavailable += 1
                } else {
                    summary.stillAvailable += 1
                }
            }
            progress(index + 1, batch.count)
        }

        let checked = summary.checked
        let gone = summary.nowUnavailable
        let left = summary.remaining
        Log.library.info("Availability sweep: checked \(checked), gone \(gone), remaining \(left)")
        return summary
    }
}

File: android/app/src/main/java/io/github/cidy02/kudos/works/WorkAvailabilitySweep.kt
package io.github.cidy02.kudos.works

import io.github.cidy02.kudos.network.ao3.AO3Clock
import io.github.cidy02.kudos.network.ao3.AO3Delay
import io.github.cidy02.kudos.network.ao3.AO3Error
import io.github.cidy02.kudos.network.ao3.AO3Result
import io.github.cidy02.kudos.network.ao3.CoroutineAO3Delay
import io.github.cidy02.kudos.network.ao3.SystemAO3Clock
import io.github.cidy02.kudos.network.ao3.work.WorkTagsRepository
import java.time.Duration
import java.time.Instant
import kotlin.coroutines.coroutineContext
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.isActive

/**
 * Manual library availability sweep — port of Apple `WorkAvailabilitySweep`.
 *
 * Deliberately **not** scheduled periodically: AO3 has no change-feed, so a
 * full-library walk is one request per work. Politeness measures match iOS:
 * 7-day recheck skip, 1500 ms serial spacing, 150-work cap, oldest-checked
 * first, and cancellation that keeps work already done.
 */
class WorkAvailabilitySweep(
    private val workRepository: WorkRepository,
    private val tagsRepository: WorkTagsRepository,
    private val clock: AO3Clock = SystemAO3Clock,
    private val delay: AO3Delay = CoroutineAO3Delay
) {
    data class Summary(
        val checked: Int = 0,
        val stillAvailable: Int = 0,
        val nowUnavailable: Int = 0,
        /** Skipped because they were checked within [RECHECK_INTERVAL]. */
        val skippedRecent: Int = 0,
        /** Eligible but beyond this run's limit, or not reached before cancellation. */
        val remaining: Int = 0,
        val cancelled: Boolean = false
    )

    companion object {
        /** Don't re-ask about a work checked more recently than this. */
        val RECHECK_INTERVAL: Duration = Duration.ofDays(7)

        /**
         * Gap between requests. Spacing goes *before* each request after the first
         * so a cancelled sweep never leaves a pointless trailing sleep.
         */
        const val REQUEST_SPACING_MILLIS: Long = 1500L

        /** Ceiling for one run so a huge library cannot become an accidental marathon. */
        const val DEFAULT_LIMIT: Int = 150
    }

    /**
     * Runs the sweep. Never throws for per-work network failure: a work that
     * cannot be checked is left as it was. Cancellation stops cleanly and
     * preserves work already done.
     */
    suspend fun sweep(limit: Int = DEFAULT_LIMIT): Summary {
        val now = Instant.ofEpochMilli(clock.nowMillis())
        val recheckCutoff = now.minus(RECHECK_INTERVAL)

        val verifiable = workRepository.listSavedWorks().filter { work ->
            work.sourceUrl.isNotBlank() &&
                !work.ao3Unavailable &&
                WorkTags.ao3WorkIdFromUrl(work.sourceUrl) != null
        }

        val pending = verifiable.filter { work ->
            val last = work.lastAvailabilityCheck
            last == null || !last.isAfter(recheckCutoff)
        }
        val skippedRecent = verifiable.size - pending.size

        // null (never checked) sorts first: those are the works we know least about.
        val ordered = pending.sortedWith(
            compareBy(nullsFirst()) { it.lastAvailabilityCheck }
        )
        val batch = ordered.take(limit)
        var remaining = ordered.size - batch.size

        var checked = 0
        var stillAvailable = 0
        var nowUnavailable = 0
        var cancelled = false

        for ((index, work) in batch.withIndex()) {
            if (!coroutineContext.isActive) {
                cancelled = true
                remaining += batch.size - index
                break
            }
            if (index > 0) {
                try {
                    delay.delay(REQUEST_SPACING_MILLIS)
                } catch (_: CancellationException) {
                    cancelled = true
                    remaining += batch.size - index
                    break
                }
            }

            val workId = WorkTags.ao3WorkIdFromUrl(work.sourceUrl) ?: continue
            when (val result = tagsRepository.refreshTags(workId)) {
                is AO3Result.Failure -> {
                    if (result.error is AO3Error.NotFound) {
                        workRepository.upsert(
                            work.copy(
                                ao3Unavailable = true,
                                lastAvailabilityCheck = now,
                                lastModifiedAt = now
                            )
                        )
                        checked++
                        nowUnavailable++
                    }
                    // Inconclusive (timeout, 5xx, parse…): leave stamps alone.
                }
                is AO3Result.Success -> {
                    workRepository.upsert(work.copy(lastAvailabilityCheck = now))
                    checked++
                    stillAvailable++
                }
            }
        }

        return Summary(
            checked = checked,
            stillAvailable = stillAvailable,
            nowUnavailable = nowUnavailable,
            skippedRecent = skippedRecent,
            remaining = remaining,
            cancelled = cancelled
        )
    }
}
